### PR TITLE
Preprocessing: Fix incrementality in Preprocessor

### LINKED ISSUE #710: Fix bugs in `QF_ALIA`
There are wrong answers to several incremental queries in benchmark `incremental/QF_ALIA/UltimateBuchiAutomizer/LexIndexValue-Pointer_true-termination.c.smt2` (sat instead of unsat).

### PR DESCRIPTION
**Pull request by blishko**
Previous implementation of handling incrementality in Preprocessor was not sound. The problem was the difference between eager push/pop and lazy application of preprocessing in MainSolver.
The lazy approach means that formulas are sent to the preprocessor only after a call to check-sat, i.e., possibly after several push commands. This meant that Preprocessor actually stored information incorrectly, because it considered all formulas processed during a check-sat command to be part of the last frame, when actually they could come from lower frames.

The proposed solution is to make tracking frames in Preprocessor lazy as well. This ensures that the frame information is stored correctly and formulas are not popped in Preprocessor when they are actually still valid in the MainSolver.

Fixes #710.

### CODE REVIEW

File: src/api/MainSolver.cc [hunk 1]
```diff
@@ -559,12 +560,32 @@ void MainSolver::Preprocessor::initialize() {
     substitutions.push();
 }
 
+void MainSolver::Preprocessor::prepareForProcessingFrame(std::size_t frameIndex) {
+    assert(frameIndex < virtualFrameCount);
+    while (actualFrameCount <= frameIndex) {
+        pushInternal();
+    }
```
Comment (Tomaqa) [inline src/api/MainSolver.cc:567]: `assert(actualFrameCount == virtualFrameCount);` ?

File: src/api/MainSolver.h [hunk 2]
```diff
@@ -172,8 +172,14 @@ class MainSolver {
         [[nodiscard]] opensmt::span<const PTRef> getPreprocessedFormulas() const;
         [[nodiscard]] Logic::SubstMap getCurrentSubstitutions() const { return substitutions.current(); }
         void setSubstitutions(std::size_t level, Logic::SubstMap && subs) { substitutions.set(level, std::move(subs)); }
+        void prepareForProcessingFrame(std::size_t frameIndex);
 
     private:
+        std::size_t virtualFrameCount {1};
+        std::size_t actualFrameCount {1};
```
Comment (Tomaqa) [inline src/api/MainSolver.h:179]: `internalFrameCount` ?

File: src/api/MainSolver.h [hunk 3]
```diff
@@ -172,8 +172,14 @@ class MainSolver {
         [[nodiscard]] opensmt::span<const PTRef> getPreprocessedFormulas() const;
         [[nodiscard]] Logic::SubstMap getCurrentSubstitutions() const { return substitutions.current(); }
         void setSubstitutions(std::size_t level, Logic::SubstMap && subs) { substitutions.set(level, std::move(subs)); }
+        void prepareForProcessingFrame(std::size_t frameIndex);
 
     private:
+        std::size_t virtualFrameCount {1};
```
Comment (Tomaqa) [inline src/api/MainSolver.h:178]: `solverFrameCount` ?

File: src/api/MainSolver.h [hunk 4]
```diff
@@ -172,8 +172,14 @@ class MainSolver {
...
+        std::size_t actualFrameCount {1};
+        void pushInternal();
+        void popInternal();
```
Comment (Tomaqa) [inline src/api/MainSolver.h:181]: ordering within a class: 1. subclasses and types, 2. member functions, 3. member variables ?

File: src/api/MainSolver.cc [hunk 5]
```diff
@@ -559,12 +560,32 @@ void MainSolver::Preprocessor::initialize() {
...
+}
 void MainSolver::Preprocessor::push() {
+    ++virtualFrameCount;
+}
+
+void MainSolver::Preprocessor::pop() {
```
Comment (Tomaqa) [inline src/api/MainSolver.cc:573]: `assert(virtualFrameCount >= actualFrameCount);` ?

File: src/api/MainSolver.cc [hunk 6]
```diff
@@ -559,12 +560,32 @@ void MainSolver::Preprocessor::initialize() {
     substitutions.push();
 }
 
+void MainSolver::Preprocessor::prepareForProcessingFrame(std::size_t frameIndex) {
+    assert(frameIndex < virtualFrameCount);
+    while (actualFrameCount <= frameIndex) {
+        pushInternal();
+    }
+}
 void MainSolver::Preprocessor::push() {
```
Comment (Tomaqa) [inline src/api/MainSolver.cc:569]: `assert(virtualFrameCount >= actualFrameCount);` ?

File: src/api/MainSolver.cc [hunk 7]
```diff
@@ -559,12 +560,32 @@ void MainSolver::Preprocessor::initialize() {
...
+    ++virtualFrameCount;
+}
+
+void MainSolver::Preprocessor::pop() {
+    --virtualFrameCount;
+    if (virtualFrameCount < actualFrameCount) {
+        assert(virtualFrameCount + 1 == actualFrameCount);
+        popInternal();
+    }
```
Comment (Tomaqa) [inline src/api/MainSolver.cc:578]: what about:
```
if (virtualFrameCount >= actualFrameCount) return;
popInternal();
assert(virtualFrameCount == actualFrameCount);
```
I think that the final state is more important than an intermediate.

File: src/api/MainSolver.cc (hunk 1 above)
Comment (blishko) [reply src/api/MainSolver.cc:567]: That does not hold, `virtualFrameCount` can still be larger.

File: src/api/MainSolver.cc (hunk 6 above)
Comment (blishko) [reply src/api/MainSolver.cc:569]: Added!

File: src/api/MainSolver.cc (hunk 5 above)
Comment (blishko) [reply src/api/MainSolver.cc:574]: Added!

File: src/api/MainSolver.h (hunk 2 above)
Comment (blishko) [reply src/api/MainSolver.h]: Renamed

File: src/api/MainSolver.h (hunk 3 above)
Comment (blishko) [reply src/api/MainSolver.h]: Renamed

File: src/api/MainSolver.h (hunk 4 above)
Comment (blishko) [reply src/api/MainSolver.h]: Fixed, if I understood it correctly.

File: src/api/MainSolver.cc (hunk 1 above)
Comment (Tomaqa) [reply src/api/MainSolver.cc:567]: Right ...

**Review by Tomaqa — APPROVED** [review]
LGTM!

**Pull request merged**